pkg/provider/provider.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/metrics/pkg/apis/custom_metrics"
+
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/provider"
+)
+
+type E2EProvider struct {
+	client dynamic.ClientPool
+	mapper apimeta.RESTMapper
+
+	values map[provider.CustomMetricInfo]int64
+}
+
+func NewE2EProvider(client dynamic.ClientPool, mapper apimeta.RESTMapper) provider.CustomMetricsProvider {
+	return &E2EProvider{
+		client: client,
+		mapper: mapper,
+		values: make(map[provider.CustomMetricInfo]int64),
+	}
+}
+
+func (p *E2EProvider) WebService() *restful.WebService {
+	ws := new(restful.WebService)
+
+	ws.Path("/metrics")
+
+	ws.Route(ws.PUT("/{namespace}/{resourceType}/{name}/{metric}/{value}").To(p.updateResource).
+		Param(ws.PathParameter("value", "value to set metric").DataType("integer").DefaultValue("0")))
+	return ws
+}
+
+func (p *E2EProvider) updateResource(request *restful.Request, response *restful.Response) {
+	namespace := request.PathParameter("namespace")
+	namespaced := false
+	if len(namespace) > 0 {
+		namespaced = true
+	}
+	resourceType := request.PathParameter("resourceType")
+	//name := request.PathParameter("name")
+	metricName := request.PathParameter("metric")
+	value := request.PathParameter("value")
+
+	groupResource := schema.GroupResource{
+		Group:    "",
+		Resource: resourceType,
+	}
+
+	info := provider.CustomMetricInfo{
+		GroupResource: groupResource,
+		Metric:        metricName,
+		Namespaced:    namespaced,
+	}
+
+	info, _, err := info.Normalized(p.mapper)
+	if err != nil {
+		glog.Errorf("Error normalizing info: %s", err)
+	}
+
+	i, err := strconv.ParseInt(value, 10, 64)
+	if err != nil {
+		glog.Errorf("Error converting value to int: %s", err)
+	}
+	p.values[info] = i
```

Comment:
you should save by `name` too.  Create a wrapper struct to store both name and info.